### PR TITLE
Bug fix: use `FULL OUTER JOIN` for dimension-only queries

### PR DESCRIPTION
**Pull request by courtneyholcomb**
### Description
Realized that we were using `LEFT OUTER JOIN` for all queries using `JoinToBaseOutputNode`. This made sense for metric queries, but not for dimension-only queries. When querying dimensions, we should use a `FULL OUTER JOIN` since we want to see all possible combinations of dimension values (including nulls). With the `LEFT OUTER JOIN` we were arbitrarily choosing one dimension's node as the source node and potentially eliminating values from all other dimensions.

### CODE REVIEW

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 1]
```diff
@@ -80,49 +80,7 @@ class DataflowRecipe:
     @property
     def join_targets(self) -> List[JoinDescription]:
         """Joins to be made to source node."""
-        join_targets = []
-        for join_recipe in self.join_linkable_instances_recipes:
```
Comment (courtneyholcomb) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:84]: Moved this logic to `JoinLinkableInstancesRecipe`, which seemed cleaner & more appropriate.

File: metricflow/dataflow/builder/node_evaluator.py [hunk 2]
```diff
@@ -71,12 +72,36 @@ class JoinLinkableInstancesRecipe:
     @property
     def join_description(self) -> JoinDescription:
```
Comment (courtneyholcomb) [inline metricflow/dataflow/builder/node_evaluator.py:73]: Prior to this PR, this attribute was not used.

File: metricflow/plan_conversion/dataflow_to_sql.py [hunk 3]
```diff
@@ -402,12 +402,9 @@ def visit_join_to_base_output_node(self, node: JoinToBaseOutputNode) -> SqlDataS
             # e.g. a data set has the dimension "listing__country_latest" and "listing" is a primary entity in the
             # data set. The next step would create an instance like "listing__listing__country_latest" without this
             # filter.
-
-            # logger.error(f"before filter is:\n{pformat_big_objects(right_data_set.instance_set.spec_set)}")
```
Comment (tlento) [inline metricflow/plan_conversion/dataflow_to_sql.py:406]: Oops. Thank you!

**Review by tlento — APPROVED** [review]
Makes sense, thank you!

**Comment by tlento** [comment]
Also if you remember please squash and merge this one - I read commit-wise at first but the final PR should be cleaner in the local git history. Thank you!

File: metricflow/dataflow/builder/node_evaluator.py [hunk 4]
```diff
@@ -306,6 +336,7 @@ def evaluate_node(
         self,
         start_node: BaseOutput,
         required_linkable_specs: Sequence[LinkableInstanceSpec],
+        default_join_type: SqlJoinType,
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/dataflow/builder/node_evaluator.py:338]: This was in preparation for my next PR - we'll need to use `CROSS JOIN` when querying `metric_time` with other dimensions (but no metrics).

[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/dataflow/builder/node_evaluator.py:338]: ... which will override the default join type for dimension-only queries.

File: metricflow/plan_conversion/sql_join_builder.py [hunk 5]
```diff
@@ -147,8 +147,21 @@ def make_base_output_join_description(
...
+            left_data_set=left_data_set, right_data_set=right_data_set, join_description=join_description
+        )
+        if join_description.join_type == SqlJoinType.CROSS_JOIN:
+            return SqlQueryPlanJoinBuilder.make_column_equality_sql_join_description(
+                right_source_node=right_data_set.data_set.sql_select_node,
+                left_source_alias=left_data_set.alias,
+                right_source_alias=right_data_set.alias,
+                column_equality_descriptions=[],
+                join_type=join_description.join_type,
+                additional_on_conditions=validity_conditions,
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/plan_conversion/sql_join_builder.py]: I'll confess I don't know anything about the validity conditions, but that behavior was not changed in this PR!

**Comment by tlento** [comment]
Makes sense, ship it!

**Pull request merged**